src/main/java/com/limechain/transaction/TransactionState.java

```diff
@@ -0,0 +1,119 @@
+package com.limechain.transaction;
+
+import com.limechain.transaction.dto.Extrinsic;
+import com.limechain.transaction.dto.ValidTransaction;
+import com.limechain.utils.ByteArrayUtils;
+import lombok.extern.java.Log;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.stereotype.Component;
+
+import java.util.Arrays;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+@Log
+@Component
+public class TransactionState {
+
+    private final TransactionPool transactionPool;
+    private final ExecutorService executor;
+    private final Queue<ValidTransaction> transactionQueue;
+
+    public TransactionState() {
+        transactionPool = new TransactionPool();
+        executor = Executors.newSingleThreadExecutor();
+        transactionQueue = new PriorityQueue<>();
+    }
+
+    public void pushTransaction(ValidTransaction validTransaction) {
+        transactionQueue.add(validTransaction);
+    }
+
+    public ValidTransaction popTransaction() {
+        return transactionQueue.poll();
+    }
+
+    public ValidTransaction popTransactionWithTimer(long timeout) throws InterruptedException {
+        ValidTransaction validTransaction = popTransaction();
+        if (validTransaction != null) return validTransaction;
+
+        Future<ValidTransaction> futureTransaction = popFutureTransaction();
+
+        try {
+            return futureTransaction.get(timeout, java.util.concurrent.TimeUnit.MILLISECONDS);
+        } catch (ExecutionException e) {
+            log.severe("Error while waiting for transaction: " + e.getMessage());
+        } catch (TimeoutException e) {
+            futureTransaction.cancel(true);
+        } catch (InterruptedException e) {
+            if (Thread.interrupted())
+                throw new InterruptedException();
+        }
```

Comment:
InterruptedException is a checked excaption and we don't want to throw it, we have ThreadInterruptedException for cases like this one